core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java

```diff
@@ -0,0 +1,127 @@
+package aima.core.search.basic.uninformed;
+
+import aima.core.search.api.BidirectionalProblem;
+import aima.core.search.api.NodeFactory;
+import aima.core.search.api.SearchController;
+import aima.core.search.api.Node;
+import aima.core.search.api.Problem;
+import aima.core.search.api.SearchForActionsFunction;
+import aima.core.search.basic.support.BasicNodeFactory;
+import aima.core.search.basic.support.BasicSearchController;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import java.util.LinkedList;
+import java.util.HashSet;
+
+/**
+ * Artificial Intelligence A Modern Approach (4th Edition): Figure ??, page ??.
+ * <br>
+ * <br>
+ *
+ * <pre>
+ * The strategy of this search implementation is inspired by the description of
+ * the bidirectional search algorithm i.e. Problem is reversed and search is performed from both
+ * frontiers.
+ * Created by manthan on 1/3/17.
+ */
+public class BidirectionalSearch<A, S> implements SearchForActionsFunction<A, S> {
+
+    @Override
+    public List<A> apply(Problem<A, S> problem) {
+        Node<A, S> node = newRootNode(problem.initialState(), 0);
+        if (problem.isGoalState(node.state())) {
+            return solution(node);
+        }
+        Node<A, S> revNode = newRootNode(((BidirectionalProblem<A, S>) problem).getReverseProblem().initialState(), 0);
```

Comment:
For now I would suggest adding a SearchForActionsBidirectionallyFunction interface similar to SearchForActionsFunction but that takes a BidirectionalProblem. Update all related code appropriately.